### PR TITLE
Adding Local skills improvement plan area as a geographic level

### PR DESCRIPTION
**Pull request by rmbielby**
Local skills improvement plan areas (LSIPs) are required by the Further Education publication team as a searchable geographic level. Explore Education Statistics now accepts LSIPs as a geographic level and this PR updates the EES data screener to match what is allowed.

The behaviour around LSIPs is:
- Local skills improvement plan area is an allowable geographic level
- Any local skills improvement plan area data rows must contain valid entries under lsip_code and lsip_name in accordance with the supplied codes and names file (lsips.csv)
- lsip_code and lsip_name are also allowed to be present in LAD geographic level rows

To accomplish the above, LSIPS are checked as part of:
- lsip_combinations()
- overcompleted_cols()

To check continued working of the screener with LSIPs, I've amended/added the following tests of the screener itself:
- Added a specific test (with failure) of the lsip_combinations() check.
- The passes_everything test checks that correct lsip entries pass all tests

### CODE REVIEW

File: data/lsips.csv [hunk 1]
```diff
@@ -0,0 +1,39 @@
+lsip_code,lsip_name
```
Comment (cjrace) [inline data/lsips.csv:1]: Would it be worth adding a section to the readme to list out data sources, could start it just with LSIPs for where we've got this from, and then do in a separate PR to add in the equivalent source information for our other lookups?

File: R/mainTests.r [hunk 2]
```diff
@@ -168,42 +169,42 @@ duplicate_rows <- function(data, meta) {
...
       filter(geographic_level != geography_matrix[15, 1]) %>%
       filter(geographic_level != geography_matrix[16, 1]) %>%
+      filter(geographic_level != geography_matrix[17, 1]) %>%
       select(all_of(present_obUnits_filters)) %>%
       get_dupes())
 
     if (nrow(dupes) > 0) {
       output <- list(
-        "message" = paste("There are", cs_num(nrow(dupes)), "duplicate rows in the data file. <br> - Note that", paste0(geography_matrix[13, 1], ", ", geography_matrix[14, 1], ", ", geography_matrix[15, 1], " and ", geography_matrix[16, 1]), "level rows were not included in this test."),
+        "message" = paste("There are", cs_num(nrow(dupes)), "duplicate rows in the data file. <br> - Note that", paste0(geography_matrix[14, 1], ", ", geography_matrix[15, 1], ", ", geography_matrix[15, 1], " and ", geography_matrix[16, 1]), "level rows were not included in this test."),
```
Comment (cjrace) [inline R/mainTests.r]: Think you missed some ID changes on this line? Has 15 twice in a row, should it be 14/15/16/17?

File: tests/testthat/test-screenFiles.R [hunk 3]
```diff
@@ -17,7 +17,7 @@ context("Passes everything")
 test_that("PassEverything", {
   screeningOutput <- testOther("../../tests/shinytest/test-data/passes_everything.csv")
 
-  expect_equal(screeningOutput$results %>% filter(!(result %in% c("PASS", "PASS WITH NOTE"))) %>% nrow(), 0)
+  expect_equal(screeningOutput$results %>% filter(!(result %in% c("PASS", "PASS WITH NOTE", "IGNORE"))) %>% nrow(), 0)
```
Comment (cjrace) [inline tests/testthat/test-screenFiles.R]: What test is giving an ignore here to make you add it to the excluded list?

File: tests/testthat/mainTests/lsip_combinations.csv
Comment (cjrace) [inline tests/testthat/mainTests/lsip_combinations.csv:1]: Can we trim this file down at all? It's noticeably bigger than our other test files 
image

File: R/mainTests.r [hunk 4]
```diff
@@ -958,6 +959,17 @@ la_col_present <- function(data) {
   return(output)
 }
 
+# Function used to check if each column for that geographic level has any cells that are not blank
```
Comment (cjrace) [inline R/mainTests.r:962]: Why is this function (col_completed) now outside of the overcomplete_regional_cols() function?

image

File: R/mainTests.r [hunk 5]
```diff
@@ -967,29 +979,18 @@ overcompleted_cols <- function(data, meta) {
 
   overcomplete_regional_cols <- function(matrixRow) {
     # Start by filtering the data down to remove the geographic level being tested and any lower levels we don't care about
```
Comment (cjrace) [inline R/mainTests.r]: Should we update or remove this comment now that we're explicitly stating the levels we care about? Maybe change to 'these are the levels that we would expect the region columns to always be empty for?'

File: R/mainTests.r [hunk 6]
```diff
@@ -967,29 +979,18 @@ overcompleted_cols <- function(data, meta) {
...
+    levels_incompatible_with_region <- c("National", "Local skills improvement plan area")
 
     level_rows <- data %>%
-      filter(geographic_level != matrixRow[1]) %>%
-      filter(!geographic_level %in% geography_matrix[3:16, ])
+      filter(geographic_level %in% levels_incompatible_with_region)
```
Comment (cjrace) [inline R/mainTests.r]: Should we still be filtering out the matrixRow level too? It was before but isn't now?

File: R/mainTests.r [hunk 7]
```diff
@@ -1001,34 +1002,44 @@ overcompleted_cols <- function(data, meta) {
     # This is a test that could benefit from more detail, and maybe a table in the error feedback
 
     # Start by filtering the data down to remove the geographic level being tested, lad rows and any lower levels we don't care about
+    levels_compatible_with_la <- c("Local authority", "Local authority district", "School", "Provider", "Institution", "Planning area")
 
     level_rows <- data %>%
-      filter(geographic_level != matrixRow[1]) %>%
-      filter(!geographic_level %in% geography_matrix[13:16, ]) %>%
-      filter(geographic_level != geography_matrix[4, 1])
+      filter(!geographic_level %in% levels_compatible_with_la)
```
Comment (cjrace) [inline R/mainTests.r]: Same question here as above, should we be filtering out the level that is being tested as well (matrixRow)?

File: R/mainTests.r (hunk 4 above)
Comment (cjrace) [reply R/mainTests.r:962]: Actually, looks like this was intentional to remove it from the scope of the function so then multiple other functions could use it and save on the copy/pasting of code?

File: R/mainTests.r [hunk 8]
```diff
@@ -1120,12 +1109,15 @@ overcompleted_cols <- function(data, meta) {
...
+  lsip_matrix <- matrix(geography_matrix[7, ], nrow = 1)
 
   overcomplete_geographies <- c(
     unlist(apply(regional_matrix, 1, overcomplete_regional_cols)),
     unlist(apply(la_matrix, 1, overcomplete_la_cols)),
-    unlist(apply(geography_matrix[4:12, ], 1, overcomplete_mid_cols)),
-    unlist(apply(geography_matrix[13:16, ], 1, overcomplete_low_cols))
+    unlist(apply(geography_matrix[4:6, ], 1, overcomplete_mid_cols)),
+    unlist(apply(lsip_matrix, 1, overcomplete_lsip_cols)),
+    unlist(apply(geography_matrix[8:13, ], 1, overcomplete_mid_cols)),
```
Comment (cjrace) [inline R/mainTests.r]: Could line 1117 be "...matrix[c(4:6, 8:13), ] ...." and remove the need for line 1119?

**Review by cjrace — CHANGES_REQUESTED** [review]

File: R/mainTests.r (hunk 4 above)
Comment (rmbielby) [reply R/mainTests.r:962]: Yep, just code neatening. I didn't like it repeating, made the code harder to read.

File: tests/testthat/test-screenFiles.R (hunk 3 above)
Comment (rmbielby) [reply tests/testthat/test-screenFiles.R]: Ah, that's a hangover from when I was figuring out what was going on. i.e. there wasn't any LSIP data in the passes everything file, so it threw an ignore for that test. I've since added an LSIP level to the passes everything file and thought I'd removed that IGNORE in the filter. Removing it for the next commit.

File: R/mainTests.r (hunk 6 above)
Comment (rmbielby) [reply R/mainTests.r]: This is filtering it out, it's only _keeping_ levels that are in the incompatible_with_region vector.

File: R/mainTests.r (hunk 7 above)
Comment (rmbielby) [reply R/mainTests.r]: Same as above, it's being filtered out using that one filter line as it's included in the levels_compatible_with_la vector.

File: R/mainTests.r (hunk 8 above)
Comment (rmbielby) [reply R/mainTests.r]: Good point, done that.

File: R/mainTests.r (hunk 5 above)
Comment (rmbielby) [reply R/mainTests.r]: Done

File: R/mainTests.r (hunk 2 above)
Comment (rmbielby) [reply R/mainTests.r]: Corrected that now

File: data/lsips.csv (hunk 1 above)
Comment (rmbielby) [reply data/lsips.csv:1]: Added that section to the readme

File: tests/testthat/mainTests/lsip_combinations.csv
Comment (rmbielby) [reply tests/testthat/mainTests/lsip_combinations.csv:1]: Brought that down to the 10s of KB level now.

File: README.html
Comment (cjrace) [inline README.html:1]: Don't need to include the html for the readme - maybe one to add to git ignore?

**Review by cjrace — CHANGES_REQUESTED** [review]

File: README.html
Comment (rmbielby) [reply README.html:1]: Done, after some fiddling around with shinytest as well.

**Review by cjrace — APPROVED** [review]
